### PR TITLE
FEATURE: Search

### PR DESCRIPTION
**Pull request by unclecheese**
Resolves: https://github.com/silverstripeltd/open-sourcerers/issues/86

## Notes

* Per ACs, spent 4 hours on an accessible autosuggest feature. It's pretty good, but may not be perfect. To turn it off, we can just comment out `search()` in the main JS file

* Reminder that this only works on the docker container. Will not work on SCP testing environment

* Will require a `docker-compose up` to run the indexing.

## Dependencies
- [ ] https://github.com/silverstripe/bambusa-theme/pull/27

### CODE REVIEW

File: app/src/Controllers/SearchController.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            ->LocaleInformation($locale);
+
+        $query = http_build_query([
+            'l' => $locale,
+            'q' => $this->getRequest()->getVar('q')
+        ]);
+
+        $link = $data->Link . '?' . $query;
+
+        $data->Link = $link;
```
Comment (chillu) [inline app/src/Controllers/SearchController.php:61]: Can you check if this is XSS?

File: app/src/PageController.php [hunk 2]
```diff
@@ -22,6 +28,37 @@ protected function init()
             Requirements::block('silverstripe/elemental-bannerblock:client/dist/styles/frontend-default.css');
         }
 
+        /**
+         * @return SearchForm
+         */
+        public function SearchForm(): SearchForm
```
Comment (chillu) [inline app/src/PageController.php:34]: 💟

File: composer.json [hunk 3]
```diff
@@ -37,7 +37,9 @@
         "wilr/silverstripe-googlesitemaps": "~2.1.5",
         "silverstripe/theme-colorpicker": "^1",
         "silverstripe/theme-fontpicker": "^1",
-        "tractorcow/silverstripe-fluent": "^4.2"
+        "tractorcow/silverstripe-fluent": "^4.2",
+        "symfony/process": "^3.2",
```
Comment (chillu) [inline composer.json:41]: What are you using `symfony/process` for?

**Review by chillu — CHANGES_REQUESTED** [review]
Lookin' good! Great job on the `search.php` workaround. A bit too much hardcoding to be general purpose, but for a sales demo it shows off what we want: Fast search.

image

Some styling issues.

Was there a reason you didn't re-enable Solr in `docker/scripts/configure-silverstripe.sh`? 

Reindex actually takes quite long on the production database locally. That's probably exacerbated by the local Docker setup being shit slow, but still not great since it delays demo creation. One of the reasons is that we're indexing everything *twice* through the built in ElementalSolrIndex. Just looking at how we can disable that thing.

File: composer.json (hunk 3 above)
Comment (unclecheese) [reply composer.json:41]: * fulltextsearch needs symfony/process ^3
* framework needs symfony/process ^whatever
* ^whatever installs 4
* fulltextsearch cannot be installed
* explicitly use previous version
* no one complains

**Comment by chillu** [comment]
I've pushed some commits which disable the double search index. Done some upstream PRs to https://github.com/dnadesign/silverstripe-elemental as well

**Review by chillu — APPROVED** [review]

**Comment by chillu** [comment]
The search results still have encoding errors, but it's good enough for now

image

File: app/src/Controllers/SearchController.php (hunk 1 above)
Comment (chillu) [reply app/src/Controllers/SearchController.php:61]: Checked a very simple one, seems fine

image

**Comment by unclecheese** [comment]
Need to reindex to resolve the encoding issue. It's an issue with the write, not the read. :)

**Pull request merged**